diff --git a/index.bun.js b/index.bun.js
--- a/index.bun.js
+++ b/index.bun.js
@@ -570,7 +570,7 @@ function _insertNode(parent, node, anchor) {
   if (anchorIndex === -1) {
     log("[INSERT]", "Could not find anchor", logId(parent), logId(anchor), "[children]", ...children.map((c) => c.id));
   }
-  parent.add(node, anchorIndex);
+  (node !== anchor || node.parent !== parent) && parent.add(node, anchorIndex);
 }
 function _removeNode(parent, node) {
   log("Removing node:", logId(node), "from parent:", logId(parent));
diff --git a/index.js b/index.js
--- a/index.js
+++ b/index.js
@@ -542,7 +542,7 @@ function _insertNode(parent, node, anchor) {
   if (anchorIndex === -1) {
     log("[INSERT]", "Could not find anchor", logId(parent), logId(anchor), "[children]", ...children.map((c) => c.id));
   }
-  parent.add(node, anchorIndex);
+  (node !== anchor || node.parent !== parent) && parent.add(node, anchorIndex);
 }
 function _removeNode(parent, node) {
   log("Removing node:", logId(node), "from parent:", logId(parent));
